File: internal/model/entity/source.go
package entity

// Source 资料来源实体
type Source struct {
	BaseEntity
	UserID          uint     `gorm:"not null;index:idx_user_notebook" json:"user_id"`     // 所属用户
	NotebookID      uint     `gorm:"not null;index:idx_user_notebook" json:"notebook_id"` // 所属笔记本
	Notebook        Notebook `gorm:"foreignKey:NotebookID;constraint:OnDelete:CASCADE"`
	Name            string   `gorm:"type:varchar(255);not null" json:"name"`                          // 来源名称
	Type            string   `gorm:"type:varchar(20);not null;index:idx_type" json:"type"`            // 类型: file/url/audio/note/youdao
	OriginalURL     string   `gorm:"type:varchar(2048)" json:"original_url"`                          // 原始URL（网址导入时）
	FilePath        string   `gorm:"type:varchar(512)" json:"file_path"`                              // 对象存储文件路径
	FileSize        int64    `json:"file_size"`                                                       // 文件大小(字节)
	MimeType        string   `gorm:"type:varchar(100)" json:"mime_type"`                              // MIME类型
	MarkdownContent string   `gorm:"type:longtext" json:"markdown_content"`                           // 解析后的Markdown内容
	Status          string   `gorm:"type:varchar(20);default:pending;index:idx_status" json:"status"` // 状态: pending/processing/ready/failed
	ErrorMessage    string   `gorm:"type:varchar(512)" json:"error_message"`                          // 失败原因
	Vectorized      bool     `gorm:"default:false" json:"vectorized"`                                 // 是否已向量化

	// 关联
	ParentBlocks []ParentBlock `gorm:"foreignKey:SourceID"`
}

func (Source) TableName() string {
	return "source"
}

File: frontend/src/api/source.ts
import client from './client';
import type { ApiResponse } from './auth';

export interface SourceData {
  id: number;
  notebook_id: number;
  name: string;
  type: 'file' | 'url' | 'audio' | 'note' | 'youdao';
  original_url: string;
  file_path: string;
  file_size: number;
  mime_type: string;
  status: 'pending' | 'processing' | 'ready' | 'failed';
  error_message: string;
  vectorized: boolean;
  created_at: string;
  updated_at: string;
}

export interface PageData<T> {
  list: T[];
  total: number;
  page: number;
  size: number;
  total_page: number;
}

// 1. List sources in a notebook (paginated)
export async function listSources(
  nbId: number,
  params?: { page?: number; size?: number; keyword?: string }
): Promise<ApiResponse<PageData<SourceData>>> {
  const res = await client.get<ApiResponse<PageData<SourceData>>>(`/notebooks/${nbId}/sources`, { params });
  return res.data;
}

// 2. Get source detail
export async function getSource(nbId: number, id: number): Promise<ApiResponse<SourceData>> {
  const res = await client.get<ApiResponse<SourceData>>(`/notebooks/${nbId}/sources/${id}`);
  return res.data;
}

// 3. Rename source
export async function renameSource(nbId: number, id: number, name: string): Promise<ApiResponse> {
  const res = await client.put<ApiResponse>(`/notebooks/${nbId}/sources/${id}`, { name });
  return res.data;
}

// 4. Delete source
export async function deleteSource(nbId: number, id: number): Promise<ApiResponse> {
  const res = await client.delete<ApiResponse>(`/notebooks/${nbId}/sources/${id}`);
  return res.data;
}

// 5. Batch delete sources
export async function batchDeleteSources(nbId: number, ids: number[]): Promise<ApiResponse> {
  const res = await client.post<ApiResponse>(`/notebooks/${nbId}/sources/batch-delete`, { ids });
  return res.data;
}

// 5.1 Delete all failed sources
export async function deleteFailedSources(nbId: number): Promise<ApiResponse<{ deleted_count: number }>> {
  const res = await client.post<ApiResponse<{ deleted_count: number }>>(`/notebooks/${nbId}/sources/delete-failed`);
  return res.data;
}

// 6. Get source markdown content
export async function getSourceContent(nbId: number, id: number): Promise<ApiResponse<{ content: string }>> {
  const res = await client.get<ApiResponse<{ content: string }>>(`/notebooks/${nbId}/sources/${id}/content`, {
    params: { _t: Date.now() },
  });
  return res.data;
}

// 7. Get source original content
export async function getSourceOriginal(nbId: number, id: number): Promise<ApiResponse<{ content: string; type: string }>> {
  const res = await client.get<ApiResponse<{ content: string; type: string }>>(`/notebooks/${nbId}/sources/${id}/original`);
  return res.data;
}

// 8. Get file download URL (presigned MinIO URL)
export async function getSourceDownloadURL(nbId: number, id: number): Promise<ApiResponse<{ url: string }>> {
  const res = await client.get<ApiResponse<{ url: string }>>(`/notebooks/${nbId}/sources/${id}/download`);
  return res.data;
}

// 9. Reimport all unvectorized sources
export async function reimportAll(): Promise<ApiResponse<{ reimported_count: number }>> {
  const res = await client.post<ApiResponse<{ reimported_count: number }>>('/sources/reimport-all');
  return res.data;
}

// 10. Reimport specific sources by IDs
export async function reimportSources(sourceIds: number[]): Promise<ApiResponse<{ reimported_count: number }>> {
  const res = await client.post<ApiResponse<{ reimported_count: number }>>('/sources/reimport', { source_ids: sourceIds });
  return res.data;
}

// 11. Save note as source
export async function createSourceFromNote(nbId: number, title: string, content: string): Promise<ApiResponse<SourceData>> {
  const res = await client.post<ApiResponse<SourceData>>(`/notebooks/${nbId}/sources/from-note`, { title, content });
  return res.data;
}

// 12. Delete source by note title
export async function deleteSourceByNote(nbId: number, title: string): Promise<ApiResponse> {
  const res = await client.post<ApiResponse>(`/notebooks/${nbId}/sources/delete-by-note`, { title });
  return res.data;
}
